File: backend/app/limiters/sliding_window_counter.py
from __future__ import annotations

import math

from redis.asyncio import Redis

from app.limiters.base import Decision


class SlidingWindowCounterLimiter:
    algorithm = "sliding_window_counter"

    async def allow(
        self,
        redis_client: Redis,
        policy_id: str,
        client_id: str,
        now_ms: int,
        params: dict[str, object],
    ) -> Decision:
        window_size_sec = int(params["window_size_sec"])
        limit = int(params["limit"])

        window_ms = window_size_sec * 1000
        curr_window_start = (now_ms // window_ms) * window_ms
        prev_window_start = curr_window_start - window_ms

        curr_key = f"rl:swc:{policy_id}:{client_id}:curr:{curr_window_start}"
        prev_key = f"rl:swc:{policy_id}:{client_id}:prev:{prev_window_start}"

        curr_raw, prev_raw = await redis_client.mget(curr_key, prev_key)
        curr_count = int(curr_raw or 0)
        prev_count = int(prev_raw or 0)

        elapsed_ms = now_ms - curr_window_start
        prev_weight = max((window_ms - elapsed_ms) / window_ms, 0)
        effective_count = curr_count + (prev_count * prev_weight)

        if effective_count < limit:
            new_count = int(await redis_client.incr(curr_key))
            await redis_client.expire(curr_key, window_size_sec * 2)
            await redis_client.expire(prev_key, window_size_sec * 2)

            remaining = max(int(math.floor(limit - (effective_count + 1))), 0)
            return Decision(
                allowed=True,
                remaining=remaining,
                algorithm_state={
                    "curr_count": new_count,
                    "prev_count": prev_count,
                    "effective_count": round(effective_count + 1, 4),
                },
            )

        retry_after_ms = max((curr_window_start + window_ms) - now_ms, 1)
        return Decision(
            allowed=False,
            reason="rate_limit_exceeded",
            remaining=0,
            retry_after_ms=retry_after_ms,
            algorithm_state={
                "curr_count": curr_count,
                "prev_count": prev_count,
                "effective_count": round(effective_count, 4),
            },
        )

File: gateway/internal/limiter/sliding_window_counter.go
package limiter

import (
	"context"
	"fmt"
	"math"
)

type SlidingWindowCounterLimiter struct {
	store RuntimeCounter
}

func NewSlidingWindowCounterLimiter(store RuntimeCounter) *SlidingWindowCounterLimiter {
	return &SlidingWindowCounterLimiter{store: store}
}

func (l *SlidingWindowCounterLimiter) Allow(ctx context.Context, policyID, clientID string, nowMS int64, params map[string]any) (Decision, error) {
	windowSizeSec, err := readPositiveInt(params, "window_size_sec")
	if err != nil {
		return Decision{}, err
	}
	limit, err := readPositiveInt(params, "limit")
	if err != nil {
		return Decision{}, err
	}

	windowMS := int64(windowSizeSec * 1000)
	currentWindowStartMS := (nowMS / windowMS) * windowMS
	previousWindowStartMS := currentWindowStartMS - windowMS

	currentKey := fmt.Sprintf("rl:sliding_window_counter:%s:%s:%d", policyID, clientID, currentWindowStartMS)
	previousKey := fmt.Sprintf("rl:sliding_window_counter:%s:%s:%d", policyID, clientID, previousWindowStartMS)

	evalResult, err := l.store.EvalSlidingWindowCounter(ctx, currentKey, previousKey, (windowSizeSec*2)+1)
	if err != nil {
		return Decision{}, err
	}

	elapsedMS := nowMS - currentWindowStartMS
	previousWeight := float64(windowMS-elapsedMS) / float64(windowMS)
	if previousWeight < 0 {
		previousWeight = 0
	}
	if previousWeight > 1 {
		previousWeight = 1
	}

	estimatedCount := float64(evalResult.CurrentWindowCount) + (float64(evalResult.PreviousWindowCount) * previousWeight)
	remaining := int(math.Floor(float64(limit) - estimatedCount))
	if remaining < 0 {
		remaining = 0
	}

	baseState := map[string]any{
		"count":                  estimatedCount,
		"window_start_ms":        currentWindowStartMS,
		"window_size_sec":        windowSizeSec,
		"state_schema_version":   1,
		"current_window_count":   evalResult.CurrentWindowCount,
		"previous_window_count":  evalResult.PreviousWindowCount,
		"previous_window_weight": previousWeight,
		"estimated_count":        estimatedCount,
	}

	if estimatedCount < float64(limit) {
		return Decision{
			Allowed:        true,
			Remaining:      &remaining,
			AlgorithmState: baseState,
		}, nil
	}

	retryAfter := int((currentWindowStartMS + windowMS) - nowMS)
	if retryAfter < 1 {
		retryAfter = 1
	}
	reason := "rate_limit_exceeded"
	zero := 0
	return Decision{
		Allowed:        false,
		Reason:         &reason,
		Remaining:      &zero,
		RetryAfterMS:   &retryAfter,
		AlgorithmState: baseState,
	}, nil
}
